src/metabase/lib/schema/expression.cljc

```diff
@@ -1,39 +1,191 @@
 (ns metabase.lib.schema.expression
   (:require
+   [malli.core :as mc]
+   [metabase.lib.schema.common :as common]
    [metabase.lib.schema.filter :as filter]
    [metabase.lib.schema.literal :as literal]
    [metabase.lib.schema.ref :as ref]
+   [metabase.lib.schema.temporal-bucketing :as temporal-bucketing]
+   [metabase.types]
    [metabase.util.malli.registry :as mr]))
 
+(comment metabase.types/keep-me)
+
+(defn- ref-of-base-type [base-type]
+  [:and
+   [:ref ::ref/ref]
+   [:fn
+    {:error/message (str ":field, :expression, :or :aggregation with a " base-type " :base-type")}
+    (fn [[_ref opts]]
+          (isa? (:base-type opts) base-type))]])
+
+(defn- ref-of-base-type-with-temporal-unit [base-type temporal-unit-schema]
```

Comment:
Tamas and I talked about this, but it's wacky and weird that date bucketing is done with a `:field` with a `:temporal-unit` rather than wrapping the `:field` in some sort of expression... as a result the expression type of a `:field` is not necessarily the same as its `:base-type`. A Field with `:base-type :type/DateTime` and `:temporal-unit :day-of-month` actually returns a `:Type/Integer`, not a `:type/DateTime`. So this lets us add that constraint in our expression typing stuff

At some point I think we should fix the `:field` clauses and have temporal bucketing be done with a separate expression; we do have `:temporal-extract` already so maybe we should just sunset `:field` `:temporal-unit` and use that everywhere instead. But I didn't want to try to fix everything all at once, so I'll open a separate issue for that stuff

Reply:
Opened https://github.com/metabase/metabase/issues/29011 to track this